Cloudfeature gsoc cp1 (#2366)
jenkins-infra/jenkins.io: content/blog/2019/07/2019-07-04-cloudfeature-gsoc-cp1.adoc

---
layout: post

> title: "GSoC Project Progress : Cloud features for external workspace manager"

Reviewer: ```suggestion
title: "GSoC Project report: Cloud features for the External Workspace Manager plugin"
```

tags:
- jenkins

> - gsoc

Reviewer: ```suggestion
- gsoc
- gsoc2019
- external-workspace-manager
```

author: yufei_zhang
description: "Short summary of coding period 1" # optional

> opengraph:

Reviewer: Either remove it entirely or add an image reference

```suggestion
```

Our project proposal was based on extending cloud features, but then we discovered we were on the wrong track.

=== Discoveries

In order to make things clear, we need to understand the true nature of EWM. As we stated earlier, we want multiple agents to share disk spaces, and the main obstacle was they all have their own workspace directory (paths) by default.
If we can modify the default workspace directory, then we might be able to utilize the shared NFS (Network File System).
And that is what EWM did.
There are other features in like different strategies choosing a disk, but the way EWM to solve the problem is simple but effective.

And by the way, from my perspective,the main problem, synchronization when visiting shared resources like two slave visit the shared file, is guaranteed by the implementation of the NFS itself, and the EWM didn't offer any synchronization mechanism to support concurrent execution.
Jenkins itself is not designed to be synchronized.

> It is left to the users to partition their jobs and let the user manage syncronization.

Reviewer 2: It is left to the users to partition their jobs and let the user manage synchronization.

Reviewer 2: fixed minor spelling typo:
syncronization -> synchronization

Reviewer 2: @martinda I don’t have permission in your branch (according to GitHub) to commit the minor typo fix suggestion.  Thus, you will have to commit my suggestion to fix the minor typo.  Thank you.

Reviewer 2: I assigned this PR to @martinda to commit:
```
fixed minor spelling typo:
syncronization -> synchronization
```

---
layout: post
title: "GSoC Project Progress : Cloud features for external workspace manager"
tags:
- jenkins
- gsoc

> author: yufei_zhang

Reviewer: This is a case for supporting multiple authors in the blogpost

Author: @oleg-nenashev I looked at the haml, but I do not know how to add that feature. It would take me a some of time.

Reviewer 2: Let’s assign to @oleg-nenashev who knows how to add multiple authors in haml

Reviewer 2: I assigned this PR to @oleg-nenashev to add haml syntax usage

description: "Short summary of coding period 1" # optional
opengraph:
---

=== About Me

> My name is Yufei Zhang, a student from Zhejiang University and Simon Fraser University Dual Degree Program. I want to share the discoveries when we are exploring the possiblities of extending current external workspace manager plugin (referred to as EWM) to support AWS EFS. 

Reviewer: Add a hyperlink to the plugin?

  vars = {
    EFS_ID        = "${aws_efs_file_system.efs_jenkins.id}"
    MOUNT_POINT   = "${var.mount_point}"
  }
}
```

As you can see, user needn't know the efs-id of the created system, just build a dependency like this, and everything is set up automatically. 

*Thirdly*, no convenience is given to user, even we implement the cloud management.
If we only implement EFS support, there are many more AWS resources, like IAM role, security groups, and the user still needs to worry about those elsewhere, and if we cannot implement all, we better implement none of them.
The point we want to address is, user better give control of AWS resources to as few softwares as possible. if resource A is managed by Jenkins, B by Ansible, C by CloudFormation, and D by Terraform, it would be a disaster to manage.

So you can see, the user can do the exactly same thing we want to implement, using other tools and current EWM, possibly in a better way.
Then we find that, though I personally did quite a lot to this idea, like learning AWS and Jenkins plugin developement and some working codes, we have to admit our idea might not provide any new benefit.

=== Done and TODOs

> This is what we've already done:

Reviewer: Link to the code which was not merged?

Author: Added links, resolving.

. The Web UI to define a AWS EFS disk
. Implemented a demo workflow code
. Implemented the AWS EFS utils, like creating, destroying, mounting an EFS.

And unfortunately we need to give up those work.
We realize sometimes negative results are also results.
A good programmer should know when he is wrong, and has the courage to admit it.
Honestly, we were a little bit uneasy when we did the decision, but it doesn't matter, we will continue and spend time on something else.

> === Future directions

Reviewer: Obsolete?
